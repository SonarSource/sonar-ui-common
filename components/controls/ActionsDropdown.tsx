/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { Link } from 'react-router';
import { LocationDescriptor } from 'history';
import { Button } from './buttons';
import Dropdown from './Dropdown';
import DropdownIcon from '../icons/DropdownIcon';
import SettingsIcon from '../icons/SettingsIcon';

interface Props {
  className?: string;
  children: React.ReactNode;
  onOpen?: () => void;
  small?: boolean;
  toggleClassName?: string;
}

export default function ActionsDropdown(props: Props) {
  return (
    <Dropdown
      className={props.className}
      onOpen={props.onOpen}
      overlay={<ul className="menu">{props.children}</ul>}>
      <Button
        className={classNames('dropdown-toggle', props.toggleClassName, {
          'button-small': props.small
        })}>
        <SettingsIcon size={props.small ? 12 : 14} />
        <DropdownIcon className="little-spacer-left" />
      </Button>
    </Dropdown>
  );
}

interface ItemProps {
  className?: string;
  children: React.ReactNode;
  destructive?: boolean;
  /** used to pass a name of downloaded file */
  download?: string;
  id?: string;
  onClick?: () => void;
  to?: LocationDescriptor;
}

export class ActionsDropdownItem extends React.PureComponent<ItemProps> {
  handleClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    if (this.props.onClick) {
      this.props.onClick();
    }
  };

  render() {
    const className = classNames(this.props.className, { 'text-danger': this.props.destructive });

    if (this.props.download && typeof this.props.to === 'string') {
      return (
        <li>
          <a
            className={className}
            download={this.props.download}
            href={this.props.to}
            id={this.props.id}>
            {this.props.children}
          </a>
        </li>
      );
    }

    if (this.props.to) {
      return (
        <li>
          <Link className={className} id={this.props.id} to={this.props.to}>
            {this.props.children}
          </Link>
        </li>
      );
    }

    return (
      <li>
        <a className={className} href="#" id={this.props.id} onClick={this.handleClick}>
          {this.props.children}
        </a>
      </li>
    );
  }
}

export function ActionsDropdownDivider() {
  return <li className="divider" />;
}
