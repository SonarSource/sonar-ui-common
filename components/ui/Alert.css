/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.alert {
  margin-bottom: var(--gridSize);
  border: 1px solid;
  border-radius: 2px;
}

.alert.is-inline {
  display: inline-flex;
}

.alert:empty {
  display: none;
}

.alert-inner {
  display: flex;
  align-items: stretch;
}

.alert-inner.is-banner {
  min-width: var(--minPageWidth);
  max-width: var(--maxPageWidth);
  margin-left: auto;
  margin-right: auto;
  padding-left: var(--pagePadding);
  padding-right: var(--pagePadding);
  box-sizing: border-box;
}

.alert-icon {
  flex: 0 0 auto;
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(4 * var(--gridSize));
  border-right: 1px solid;
}

.alert-icon.is-banner {
  width: calc(2 * var(--gridSize));
  border-right: none;
}

.alert-content {
  flex: 1 0 0;
  padding: var(--gridSize) calc(2 * var(--gridSize));
  overflow: auto;
  text-align: left;
}

.alert-content a,
.alert-content .button-link {
  border-color: var(--darkBlue);
}

.alert-error {
  border-color: var(--alertBorderError);
  background-color: var(--alertBackgroundError);
  color: var(--alertTextError);
}

.alert-error .alert-icon {
  border-color: var(--alertBorderError);
}

.alert-warning {
  border-color: var(--alertBorderWarning);
  background-color: var(--alertBackgroundWarning);
  color: var(--alertTextWarning);
}

.alert-warning .alert-icon {
  border-color: var(--alertBorderWarning);
}

.alert-info {
  border-color: var(--alertBorderInfo);
  background-color: var(--alertBackgroundInfo);
  color: var(--alertTextInfo);
}

.alert-info .alert-icon {
  border-color: var(--alertBorderInfo);
}

.alert-success {
  border-color: var(--alertBorderSuccess);
  background-color: var(--alertBackgroundSuccess);
  color: var(--alertTextSuccess);
}

.alert-success .alert-icon {
  border-color: var(--alertBorderSuccess);
}
