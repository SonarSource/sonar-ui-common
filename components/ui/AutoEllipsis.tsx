/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import classNames from 'classnames';
import * as React from 'react';

type EllipsisPredicate = (
  node: HTMLElement,
  props: Omit<AutoEllipsisProps, 'customShouldEllipsis'>
) => boolean;

interface AutoEllipsisProps {
  customShouldEllipsis?: EllipsisPredicate;
  maxHeight?: number;
  maxWidth?: number;
  useParent?: boolean;
}

interface Props extends AutoEllipsisProps {
  children: React.ReactElement;
}

/*
 * This component allows to automatically add the .text-ellipsis class on it's children if this one
 * might overflow the max width/height passed as props.
 * If one of maxHeight or maxWidth is not specified, they will be ignored in the conditions to add the ellipsis class.
 * If useParent is true, then the parent size will be used instead of the undefined maxHeight/maxWidth
 */
export default function AutoEllipsis(props: Props) {
  const { children, ...autoEllipsisProps } = props;
  const [autoEllispis, ref] = useAutoEllipsis(autoEllipsisProps);

  return React.cloneElement(children, {
    className: classNames(children.props.className, { 'text-ellipsis': autoEllispis }),
    ref,
  });
}

export function useAutoEllipsis(props: AutoEllipsisProps): [boolean, (node: HTMLElement) => void] {
  const [autoEllipsis, setAutoEllipsis] = React.useState(false);

  // useCallback instead of useRef to be able to compute if the flag is needed as soon as the ref is attached
  // useRef doesn't accept a callback to notify us that the current ref value was attached,
  // see https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node for more info on this.
  const ref = React.useCallback(
    (node: HTMLElement) => {
      if (!autoEllipsis && node) {
        const shouldEllipsis = props.customShouldEllipsis ?? defaultShouldEllipsis;
        setAutoEllipsis(shouldEllipsis(node, props));
      }
    },
    // We don't want to apply this effect when ellipsis state change, only this effect can change it
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [props.customShouldEllipsis, props.maxHeight, props.maxWidth, props.useParent]
  );

  return [autoEllipsis, ref];
}

export const defaultShouldEllipsis: EllipsisPredicate = (
  node,
  { useParent = true, maxWidth, maxHeight }
) => {
  if (node.parentElement && useParent) {
    maxWidth = maxWidth ?? node.parentElement.clientWidth;
    maxHeight = maxHeight ?? node.parentElement.clientHeight;
  }
  return (
    (maxWidth !== undefined && node.clientWidth > maxWidth) ||
    (maxHeight !== undefined && node.clientHeight > maxHeight)
  );
};
