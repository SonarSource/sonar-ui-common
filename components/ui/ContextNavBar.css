/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.navbar-context,
.navbar-context .navbar-inner {
  background-color: #fff;
  z-index: var(--contextbarZIndex);
}

.navbar-context .navbar-inner {
  padding-top: var(--gridSize);
  border-bottom: 1px solid var(--barBorderColor);
}

.navbar-context .navbar-inner-with-notif {
  border-bottom: none;
}

.navbar-context-justified {
  display: flex;
  justify-content: space-between;
}

/* use `min-width: 0` to cut breadcrumb links (to end with "...") */
/* https://stackoverflow.com/questions/38223879/white-space-nowrap-breaks-flexbox-layout */
.navbar-context-header {
  display: flex;
  align-items: center;
  min-width: 0;
  height: calc(4 * var(--gridSize));
  font-size: var(--bigFontSize);
}

/* disallow icons and slash separators to shrink */
.navbar-context-header > *:not(.navbar-context-header-breadcrumb-link) {
  flex-shrink: 0;
}

.navbar-context-header-breadcrumb-link {
  min-width: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.navbar-context-header .slash-separator {
  margin-left: var(--gridSize);
  margin-right: var(--gridSize);
  font-size: 24px;
}

.navbar-context-header .slash-separator::after {
  color: rgba(68, 68, 68, 0.2);
}

/* set `min-width: 0` to allow flexbox item to shrink */
/* https://stackoverflow.com/questions/38223879/white-space-nowrap-breaks-flexbox-layout */
.navbar-context-meta {
  display: flex;
  align-items: center;
  height: calc(4 * var(--gridSize));
  padding-left: 20px;
  color: var(--secondFontColor);
  font-size: var(--smallFontSize);
  text-align: right;
}

.navbar-context-meta-secondary {
  position: absolute;
  top: 34px;
  right: 0;
  padding: 0 20px;
  white-space: nowrap;
}

.navbar-context-description {
  display: inline-block;
  line-height: var(--controlHeight);
  margin-left: var(--gridSize);
  padding-top: 4px;
  color: var(--secondFontColor);
  font-size: var(--smallFontSize);
}
