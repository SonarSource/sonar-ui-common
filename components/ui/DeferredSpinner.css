/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.deferred-spinner {
  position: relative;
  vertical-align: middle;
  width: 16px;
  height: 16px;
  border: 2px solid var(--blue);
  border-radius: 50%;
  animation: spin 0.75s infinite linear;
}

.deferred-spinner:before,
.deferred-spinner:after {
  left: -2px;
  top: -2px;
  display: none;
  position: absolute;
  content: '';
  width: inherit;
  height: inherit;
  border: inherit;
  border-radius: inherit;
}

.deferred-spinner,
.deferred-spinner:before,
.deferred-spinner:after {
  display: inline-block;
  box-sizing: border-box;
  border-color: transparent;
  border-top-color: var(--blue);
  animation-duration: 1.2s;
}

.deferred-spinner:before {
  transform: rotate(120deg);
}

.deferred-spinner:after {
  transform: rotate(240deg);
}

.deferred-spinner-placeholder {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  width: 16px;
  height: 16px;
  visibility: hidden;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}
