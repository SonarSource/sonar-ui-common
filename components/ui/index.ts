/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './Alert';
export * from './ContextNavBar';
export * from './DeferredSpinner';
export * from './DuplicationsRating';
export * from './FilesCounter';
export * from './GenericAvatar';
export * from './Level';
export * from './NavBar';
export * from './NavBarTabs';
export * from './NewsBox';
export * from './PageActions';
export * from './popups';
export * from './Rating';
export * from './SizeRating';
export * from './update-center';
