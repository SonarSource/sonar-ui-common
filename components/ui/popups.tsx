/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import ClickEventBoundary from '../controls/ClickEventBoundary';
import './popups.css';

export enum PopupPlacement {
  Bottom = 'bottom',
  BottomLeft = 'bottom-left',
  BottomRight = 'bottom-right',
  LeftTop = 'left-top',
  RightTop = 'right-top'
}

interface PopupProps {
  arrowStyle?: React.CSSProperties;
  children?: React.ReactNode;
  className?: string;
  noPadding?: boolean;
  placement?: PopupPlacement;
  style?: React.CSSProperties;
}

export function Popup(props: PopupProps) {
  const { placement = PopupPlacement.Bottom } = props;
  return (
    <ClickEventBoundary>
      <div
        className={classNames(
          'popup',
          `is-${placement}`,
          { 'no-padding': props.noPadding },
          props.className
        )}
        style={props.style}>
        {props.children}
        <PopupArrow style={props.arrowStyle} />
      </div>
    </ClickEventBoundary>
  );
}

interface PopupArrowProps {
  style?: React.CSSProperties;
}

export function PopupArrow(props: PopupArrowProps) {
  return <div className="popup-arrow" style={props.style} />;
}
