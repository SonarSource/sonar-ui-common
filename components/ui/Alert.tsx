/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import { translate } from '../../helpers/l10n';
import { Dict } from '../../types/types';
import AlertErrorIcon from '../icons/AlertErrorIcon';
import AlertSuccessIcon from '../icons/AlertSuccessIcon';
import AlertWarnIcon from '../icons/AlertWarnIcon';
import InfoIcon from '../icons/InfoIcon';
import { css, styled, Theme, ThemedProps, useTheme } from '../theme';
import DeferredSpinner from './DeferredSpinner';

type AlertDisplay = 'banner' | 'inline' | 'block';
type AlertVariant = 'error' | 'warning' | 'success' | 'info' | 'loading';

export interface AlertProps {
  display?: AlertDisplay;
  variant: AlertVariant;
}

interface AlertVariantInformation {
  icon: JSX.Element;
  color: string;
  borderColor: string;
  backGroundColor: string;
}

const StyledAlertIcon = styled.div<{ isBanner: boolean; variantInfo: AlertVariantInformation }>`
  flex: 0 0 auto;
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(${(props) => (props.isBanner ? 2 : 4)} * ${(props) => props.theme.sizes.gridSize});
  border-right: ${(props) => (!props.isBanner ? '1px solid' : 'none')};
  border-color: ${(props) => props.variantInfo.borderColor};
`;

const StyledAlertContent = styled.div`
  flex: 1 1 auto;
  overflow: auto;
  text-align: left;
  padding: ${(props) => props.theme.sizes.gridSize}
    calc(2 * ${(props) => props.theme.sizes.gridSize});
`;

const alertInnerIsBannerMixin = ({ theme }: ThemedProps) => css`
  min-width: ${theme.sizes.minPageWidth};
  max-width: ${theme.sizes.maxPageWidth};
  margin-left: auto;
  margin-right: auto;
  padding-left: ${theme.sizes.pagePadding};
  padding-right: ${theme.sizes.pagePadding};
  box-sizing: border-box;
`;

const StyledAlertInner = styled.div<{ isBanner: boolean }>`
  display: flex;
  align-items: stretch;
  ${(props) => (props.isBanner ? alertInnerIsBannerMixin : null)}
`;

const StyledAlert = styled.div<{ isInline: boolean; variantInfo: AlertVariantInformation }>`
  border: 1px solid;
  border-radius: 2px;
  margin-bottom: ${(props) => props.theme.sizes.gridSize};
  border-color: ${(props) => props.variantInfo.borderColor};
  background-color: ${(props) => props.variantInfo.backGroundColor};
  color: ${(props) => props.variantInfo.color};
  display: ${(props) => (props.isInline ? 'inline-block' : 'block')};

  :empty {
    display: none;
  }

  a,
  .button-link {
    border-color: ${(props) => props.theme.colors.darkBlue};
  }
`;

function getAlertVariantInfo({ colors }: Theme, variant: AlertVariant): AlertVariantInformation {
  const variantList: Dict<AlertVariantInformation> = {
    error: {
      icon: <AlertErrorIcon fill={colors.alertIconError} />,
      color: colors.alertTextError,
      borderColor: colors.alertBorderError,
      backGroundColor: colors.alertBackgroundError,
    },
    warning: {
      icon: <AlertWarnIcon fill={colors.alertIconWarning} />,
      color: colors.alertTextWarning,
      borderColor: colors.alertBorderWarning,
      backGroundColor: colors.alertBackgroundWarning,
    },
    success: {
      icon: <AlertSuccessIcon fill={colors.alertIconSuccess} />,
      color: colors.alertTextSuccess,
      borderColor: colors.alertBorderSuccess,
      backGroundColor: colors.alertBackgroundSuccess,
    },
    info: {
      icon: <InfoIcon fill={colors.alertIconInfo} />,
      color: colors.alertTextInfo,
      borderColor: colors.alertBorderInfo,
      backGroundColor: colors.alertBackgroundInfo,
    },
    loading: {
      icon: <DeferredSpinner timeout={0} />,
      color: colors.alertTextInfo,
      borderColor: colors.alertBorderInfo,
      backGroundColor: colors.alertBackgroundInfo,
    },
  };

  return variantList[variant];
}

export function Alert(props: AlertProps & React.HTMLAttributes<HTMLDivElement>) {
  const theme = useTheme();
  const { className, display, variant, ...domProps } = props;
  const isInline = display === 'inline';
  const isBanner = display === 'banner';
  const variantInfo = getAlertVariantInfo(theme, variant);

  return (
    <StyledAlert
      className={classNames('alert', className)}
      isInline={isInline}
      role="alert"
      variantInfo={variantInfo}
      {...domProps}>
      <StyledAlertInner isBanner={isBanner}>
        <StyledAlertIcon
          aria-label={translate('alert.tooltip', variant)}
          isBanner={isBanner}
          variantInfo={variantInfo}>
          {variantInfo.icon}
        </StyledAlertIcon>
        <StyledAlertContent className="alert-content">{props.children}</StyledAlertContent>
      </StyledAlertInner>
    </StyledAlert>
  );
}
