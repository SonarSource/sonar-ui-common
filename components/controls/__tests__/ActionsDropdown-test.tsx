/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable sonarjs/no-duplicate-string */
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { click } from '../../../helpers/testUtils';
import { PopupPlacement } from '../../ui/popups';
import ActionsDropdown, {
  ActionsDropdownDivider,
  ActionsDropdownItem,
  ActionsDropdownProps,
} from '../ActionsDropdown';

describe('ActionsDropdown', () => {
  it('should render correctly', () => {
    expect(shallowRender()).toMatchSnapshot();
    expect(shallowRender({ small: false })).toMatchSnapshot();
  });

  function shallowRender(props: Partial<ActionsDropdownProps> = {}) {
    return shallow(
      <ActionsDropdown
        className="foo"
        onOpen={jest.fn()}
        overlayPlacement={PopupPlacement.Bottom}
        small={true}
        toggleClassName="bar"
        {...props}>
        <span>Hello world</span>
      </ActionsDropdown>
    );
  }
});

describe('ActionsDropdownItem', () => {
  it('should render correctly', () => {
    expect(shallowRender()).toMatchSnapshot();
    expect(shallowRender({ destructive: true, id: 'baz', to: 'path/name' })).toMatchSnapshot();
    expect(shallowRender({ download: 'foo/bar', to: 'path/name' })).toMatchSnapshot();
  });

  it('should trigger click', () => {
    const onClick = jest.fn();
    const wrapper = shallowRender({ onClick });
    click(wrapper.find('a'));
    expect(onClick).toBeCalled();
  });

  it('should render correctly copy item', () => {
    const wrapper = mountRender({ copyValue: 'my content to copy to clipboard' });
    expect(wrapper).toMatchSnapshot();
  });

  function shallowRender(props: Partial<ActionsDropdownItem['props']> = {}) {
    return shallow(renderContent(props));
  }

  function mountRender(props: Partial<ActionsDropdownItem['props']> = {}) {
    return mount(renderContent(props));
  }

  function renderContent(props: Partial<ActionsDropdownItem['props']> = {}) {
    return (
      <ActionsDropdownItem className="foo" {...props}>
        <span>Hello world</span>
      </ActionsDropdownItem>
    );
  }
});

describe('ActionsDropdownDivider', () => {
  it('should render correctly', () => {
    expect(shallow(<ActionsDropdownDivider />)).toMatchSnapshot();
  });
});
