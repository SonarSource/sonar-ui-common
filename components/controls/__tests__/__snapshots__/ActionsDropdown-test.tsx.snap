// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsDropdown should render correctly 1`] = `
<Dropdown
  className="foo"
  onOpen={[MockFunction]}
  overlay={
    <ul
      className="menu"
    >
      <span>
        Hello world
      </span>
    </ul>
  }
  overlayPlacement="bottom"
>
  <Button
    className="dropdown-toggle bar button-small"
  >
    <SettingsIcon
      size={12}
    />
    <DropdownIcon
      className="little-spacer-left"
    />
  </Button>
</Dropdown>
`;

exports[`ActionsDropdown should render correctly 2`] = `
<Dropdown
  className="foo"
  onOpen={[MockFunction]}
  overlay={
    <ul
      className="menu"
    >
      <span>
        Hello world
      </span>
    </ul>
  }
  overlayPlacement="bottom"
>
  <Button
    className="dropdown-toggle bar"
  >
    <SettingsIcon
      size={14}
    />
    <DropdownIcon
      className="little-spacer-left"
    />
  </Button>
</Dropdown>
`;

exports[`ActionsDropdownDivider should render correctly 1`] = `
<li
  className="divider"
/>
`;

exports[`ActionsDropdownItem should render correctly 1`] = `
<li>
  <a
    className="foo"
    href="#"
    onClick={[Function]}
  >
    <span>
      Hello world
    </span>
  </a>
</li>
`;

exports[`ActionsDropdownItem should render correctly 2`] = `
<li>
  <Link
    className="foo text-danger"
    id="baz"
    onlyActiveOnIndex={false}
    style={Object {}}
    to="path/name"
  >
    <span>
      Hello world
    </span>
  </Link>
</li>
`;

exports[`ActionsDropdownItem should render correctly 3`] = `
<li>
  <a
    className="foo"
    download="foo/bar"
    href="path/name"
  >
    <span>
      Hello world
    </span>
  </a>
</li>
`;
